### PR TITLE
ocamlPackages: update a bunch of libraries from the mirage universe

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Motivation for this change

* `Cstruct.len` has been finally deprecated in ctruct 6.0.1. A lot of libraries will have to be updated for this (see <https://github.com/ocaml/opam-repository/pull/19143>, <https://github.com/ocaml/opam-repository/pull/19142>).
* `tls` 0.14 breaks `paf`
* `paf` has a new release which supports `tls` 0.14 which breaks `git-paf` ~~which hasn't had a new release yet.~~
* `paf` >= 0.0.4 needs `letsencrypt` 0.3.0

More to come most likely, we are now waiting for a lot of `mirage-*` libraries to update as well as things like `conduit-lwt-unix`.

One thing to watch is also the use of `io-page-unix`. If nothing depends on it in nixpkgs anymore we can update `io-page`.

cc @vbgl for their information

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/ocaml-modules/cstruct/ppx.nix [hunk 1]
```diff
@@ -13,20 +13,8 @@ buildDunePackage {
 
   minimumOCamlVersion = "4.07";
 
-  # prevent ANSI escape sequences from messing up the test cases
-  # https://github.com/mirage/ocaml-cstruct/issues/283
-  patches = [
-    (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/development/ocaml-modules/cstruct/ppx.nix:19]: Please remove the unused fetchpatch from inputs.

**Comment by sternenseemann** [comment]
Still failing:

- [ ] jackline
- [ ] ocamlPackages.mirage-block-combinators
- [ ] ocamlPackages.mirage-block-ramdisk
- [ ] ocamlPackages.tezos-error-monad
- [ ] ocamlPackages.tezos-micheline
- [ ] ocamlPackages.tezos-protocol-environment-sigs
- [ ] ocamlPackages.tezos-rpc
- [ ] ocamlPackages.tezos-stdlib
- [ ] ocamlPackages.tls-async

**Comment by tg-x** [comment]
#139644 also updates tls, i suggest closing that one and adding the x509 and macaddr update to this PR

**Comment by sternenseemann** [comment]
Can try that later when I get home again, sounds good. Thanks for the heads up!

**Comment by sternenseemann** [comment]
`nixpkgs-review` is clean now.

**Comment by tg-x** [comment]
OK, so looks like we could update x509 to 0.14.1 for now

File: pkgs/development/ocaml-modules/cstruct/ppx.nix [hunk 2]
```diff
@@ -13,20 +12,8 @@ buildDunePackage {
 
   minimumOCamlVersion = "4.07";
```
Comment (vbgl) [inline pkgs/development/ocaml-modules/cstruct/ppx.nix:13]:
```suggestion
  minimalOCamlVersion = "4.08";
```

Build of `ocaml-ng.ocamlPackages_4_07.ppx_cstruct` fails with:

>   File "ppx/ppx_cstruct.ml", line 347, characters 27-36:
  Error: This expression has type 'a option
         but an expression was expected of type string

Comment (sternenseemann) [reply pkgs/development/ocaml-modules/cstruct/ppx.nix]: Done.

Comment (vbgl) [reply pkgs/development/ocaml-modules/cstruct/ppx.nix]: This was also a suggestion to fix the spelling…

**Comment by vbgl** [comment]
Is there any value in the git history of this PR or can I squash it into a single commit (I did not test the intermediate states but I guess that there may be some broken packages in them)?

**Review by vbgl — APPROVED** [review]

File: pkgs/development/ocaml-modules/cstruct/ppx.nix (hunk 2 above)
Comment (sternenseemann) [reply pkgs/development/ocaml-modules/cstruct/ppx.nix]: Oh right, completely forgot that change happened. Maybe we should remove those treewide before 21.11 and make the old spelling a warning? Either way fixed.

**Comment by sternenseemann** [comment]
> Is there any value in the git history of this PR or can I squash it into a single commit (I did not test the intermediate states but I guess that there may be some broken packages in them)?

I believe a merge commit should make them atomic? But yes, there is most likely intermediate breakage.

**Pull request merged**